File: java/src/main/java/com/ibm/was/dashboard/service/WebSphereClient.java
package com.ibm.was.dashboard.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.Map;

/**
 * Manages WebSphere / ODR / CPE / ICN servers via SSH.
 *
 * Uses JSch to SSH into the node and run:
 *   $WAS_HOME/profiles/$PROFILE/bin/startServer.sh  $SERVER_NAME
 *   $WAS_HOME/profiles/$PROFILE/bin/stopServer.sh   $SERVER_NAME
 *
 * Status is determined by TCP port probe (no special WAS API needed).
 */
@Component
public class WebSphereClient {

    private static final Logger log = LoggerFactory.getLogger(WebSphereClient.class);
    private static final int TCP_TIMEOUT_MS = 5_000;

    // ── Status ──────────────────────────────────────────────────────

    /**
     * Returns ["running"|"stopped"|"unknown", message].
     */
    public String[] getStatus(Map<String, Object> server) {
        String host = str(server, "host");
        if (host.isEmpty()) return new String[]{"unknown", "No host configured"};

        Integer httpPort  = intVal(server, "http_port");
        Integer httpsPort = intVal(server, "https_port");

        for (Integer port : new Integer[]{httpPort, httpsPort}) {
            if (port != null && portOpen(host, port)) {
                return new String[]{"running", "Port " + port + " open on " + host};
            }
        }

        if (httpPort == null && httpsPort == null) {
            return new String[]{"unknown", "No ports configured"};
        }
        return new String[]{"stopped", "No open port found on " + host};
    }

    public static boolean portOpen(String host, int port) {
        try (Socket s = new Socket()) {
            s.connect(new InetSocketAddress(host, port), TCP_TIMEOUT_MS);
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    // ── SSH actions ──────────────────────────────────────────────────

    public String[] start(Map<String, Object> server) {
        return sshAction(server, "startServer");
    }

    public String[] stop(Map<String, Object> server) {
        return sshAction(server, "stopServer");
    }

    public String[] restart(Map<String, Object> server) {
        String[] result = stop(server);
        if (!result[0].equals("ok")) return result;
        return start(server);
    }

    private String[] sshAction(Map<String, Object> server, String script) {
        String serverName = str(server, "server_name");
        if (serverName.isEmpty()) return new String[]{"error", "server_name not configured"};

        String wasHome  = str(server, "was_home");
        String profile  = str(server, "profile_name");
        String adminUser = str(server, "admin_username");
        String adminPass = str(server, "admin_password");

        String cmd = wasHome + "/profiles/" + profile + "/bin/" + script + ".sh " + serverName;
        if (script.equals("stopServer") && !adminPass.isEmpty()) {
            cmd += " -username " + adminUser + " -password " + adminPass;
        }

        try {
            String output = execSsh(server, cmd, 180);
            boolean ok = output.contains("ADMU3000I") || output.contains("ADMU4000I")
                      || output.contains("open for e-business")
                      || output.contains("stop completed");
            if (!ok && !output.contains("error") && !output.contains("Error")) ok = true;
            return new String[]{ok ? "ok" : "error",
                                (ok ? script + " succeeded: " : script + " failed: ") + truncate(output, 300)};
        } catch (Exception e) {
            log.error("SSH {} failed for {}: {}", script, server.get("host"), e.getMessage());
            return new String[]{"error", "SSH error: " + e.getMessage()};
        }
    }

    // ── JSch SSH execution ─────────────────────────────────────────

    private String execSsh(Map<String, Object> server, String command, int timeoutSec) throws Exception {
        try {
            Class.forName("com.jcraft.jsch.JSch");
        } catch (ClassNotFoundException e) {
            return "[SIMULATION] JSch not available. Command would be: " + command;
        }

        com.jcraft.jsch.JSch jsch = new com.jcraft.jsch.JSch();

        String sshUser    = str(server, "ssh_username");
        String sshKeyPath = str(server, "ssh_key");
        String sshPass    = str(server, "ssh_password");
        String host       = str(server, "host");

        if (!sshKeyPath.isEmpty()) {
            jsch.addIdentity(sshKeyPath);
        }

        com.jcraft.jsch.Session session = jsch.getSession(sshUser, host, 22);
        session.setConfig("StrictHostKeyChecking", "no");
        if (!sshPass.isEmpty()) session.setPassword(sshPass);
        session.setTimeout(timeoutSec * 1000);
        session.connect(timeoutSec * 1000);

        try {
            com.jcraft.jsch.ChannelExec channel =
                (com.jcraft.jsch.ChannelExec) session.openChannel("exec");
            channel.setCommand(command);

            InputStream stdout = channel.getInputStream();
            InputStream stderr = channel.getErrStream();
            channel.connect();

            StringBuilder sb = new StringBuilder();
            byte[] buf = new byte[4096];
            int n;
            while ((n = stdout.read(buf)) != -1)  sb.append(new String(buf, 0, n));
            while ((n = stderr.read(buf))  != -1) sb.append(new String(buf, 0, n));

            channel.disconnect();
            return sb.toString();
        } finally {
            session.disconnect();
        }
    }

    // ── Helpers ───────────────────────────────────────────────────

    private static String str(Map<String, Object> m, String key) {
        Object v = m.get(key);
        return v == null ? "" : v.toString();
    }

    private static Integer intVal(Map<String, Object> m, String key) {
        Object v = m.get(key);
        if (v instanceof Number n) return n.intValue();
        if (v instanceof String s && !s.isEmpty()) {
            try { return Integer.parseInt(s); } catch (NumberFormatException ignored) {}
        }
        return null;
    }

    private static String truncate(String s, int max) {
        return s.length() > max ? s.substring(0, max) + "…" : s;
    }
}

File: backend/websphere_client.py
"""
WebSphere SSH client.

Uses paramiko to SSH into a WAS node and execute:
  - startServer.sh / stopServer.sh  (start/stop)
  - TCP port check                  (status)

All operations are wrapped so they never raise – they return
(success: bool, message: str) tuples.
"""
import socket
import logging
import os
from typing import Tuple

logger = logging.getLogger(__name__)

try:
    import paramiko
    PARAMIKO_AVAILABLE = True
except ImportError:
    PARAMIKO_AVAILABLE = False
    logger.warning("paramiko not installed – SSH management disabled. Run: pip install paramiko")


def check_port(host: str, port: int, timeout: float = 5.0) -> bool:
    """Return True if TCP port is open (server is likely running)."""
    try:
        with socket.create_connection((host, port), timeout=timeout):
            return True
    except (OSError, socket.timeout):
        return False


def _get_ssh_client(server: dict) -> "paramiko.SSHClient":
    """Build and connect a paramiko SSH client from server config."""
    if not PARAMIKO_AVAILABLE:
        raise RuntimeError("paramiko is not installed")

    ssh = paramiko.SSHClient()
    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())  # enterprise network

    key_path = server.get("ssh_key")
    username = server.get("ssh_username", "wasadmin")
    password = server.get("ssh_password", "")
    host = server["host"]

    connect_kwargs = {"username": username, "timeout": 30}
    if key_path and os.path.exists(key_path):
        connect_kwargs["key_filename"] = key_path
    elif password:
        connect_kwargs["password"] = password
    else:
        logger.warning("No SSH credential provided for %s – trying key-based auth", host)

    ssh.connect(host, **connect_kwargs)
    return ssh


def _ssh_exec(server: dict, command: str, timeout: int = 120) -> Tuple[int, str, str]:
    """Execute a command over SSH. Returns (exit_code, stdout, stderr)."""
    ssh = _get_ssh_client(server)
    try:
        _, stdout, stderr = ssh.exec_command(command, timeout=timeout)
        exit_code = stdout.channel.recv_exit_status()
        return exit_code, stdout.read().decode(), stderr.read().decode()
    finally:
        ssh.close()


# ── Public API ────────────────────────────────────────────────────

def get_server_status(server: dict) -> Tuple[str, str]:
    """
    Returns (status_string, message).
    status_string: 'running' | 'stopped' | 'unknown'
    """
    http_port = server.get("http_port")
    https_port = server.get("https_port")
    host = server.get("host", "")

    if not host:
        return "unknown", "No host configured"

    # Prefer HTTP port check; fall back to HTTPS
    for port in filter(None, [http_port, https_port]):
        if check_port(host, port):
            return "running", f"Port {port} is open on {host}"

    # If no port configured at all, report unknown
    if not http_port and not https_port:
        return "unknown", "No port configured for status check"

    return "stopped", f"No open port found on {host}"


def start_server(server: dict) -> Tuple[bool, str]:
    """SSH to the node and run startServer.sh."""
    if not PARAMIKO_AVAILABLE:
        return False, "paramiko not installed – cannot manage servers via SSH"

    was_home = server.get("was_home", "/opt/IBM/WebSphere/AppServer")
    profile = server.get("profile_name", "AppSrv01")
    server_name = server.get("server_name", "")
    if not server_name:
        return False, "server_name not configured"

    cmd = f"{was_home}/profiles/{profile}/bin/startServer.sh {server_name}"
    logger.info("Starting WAS server: %s on %s", server_name, server["host"])
    try:
        rc, stdout, stderr = _ssh_exec(server, cmd, timeout=180)
        output = stdout + stderr
        if rc == 0 or "ADMU3000I" in output or "open for e-business" in output:
            return True, f"Server {server_name} started successfully"
        return False, f"startServer returned {rc}: {output[:500]}"
    except Exception as exc:
        logger.exception("start_server failed for %s", server_name)
        return False, f"SSH error: {exc}"


def stop_server(server: dict) -> Tuple[bool, str]:
    """SSH to the node and run stopServer.sh."""
    if not PARAMIKO_AVAILABLE:
        return False, "paramiko not installed – cannot manage servers via SSH"

    was_home = server.get("was_home", "/opt/IBM/WebSphere/AppServer")
    profile = server.get("profile_name", "AppSrv01")
    server_name = server.get("server_name", "")
    admin_user = server.get("admin_username", "wsadmin")
    admin_pass = server.get("admin_password", "")

    if not server_name:
        return False, "server_name not configured"

    auth_flags = ""
    if admin_pass:
        auth_flags = f"-username {admin_user} -password {admin_pass}"

    cmd = f"{was_home}/profiles/{profile}/bin/stopServer.sh {server_name} {auth_flags}"
    logger.info("Stopping WAS server: %s on %s", server_name, server["host"])
    try:
        rc, stdout, stderr = _ssh_exec(server, cmd, timeout=180)
        output = stdout + stderr
        if rc == 0 or "ADMU4000I" in output or "stop completed" in output.lower():
            return True, f"Server {server_name} stopped successfully"
        return False, f"stopServer returned {rc}: {output[:500]}"
    except Exception as exc:
        logger.exception("stop_server failed for %s", server_name)
        return False, f"SSH error: {exc}"


def restart_server(server: dict) -> Tuple[bool, str]:
    """Stop then start."""
    ok, msg = stop_server(server)
    if not ok:
        return False, f"Stop failed: {msg}"
    ok, msg = start_server(server)
    if not ok:
        return False, f"Start failed: {msg}"
    return True, f"Server {server.get('server_name')} restarted successfully"
